test/access/Ownable.test.js

```diff
@@ -1,72 +1,80 @@
-const { constants, expectEvent } = require('@openzeppelin/test-helpers');
-const { expectRevertCustomError } = require('../helpers/customError');
-
-const { ZERO_ADDRESS } = constants;
-
+const { ethers } = require('hardhat');
 const { expect } = require('chai');
-
-const Ownable = artifacts.require('$Ownable');
-
-contract('Ownable', function (accounts) {
-  const [owner, other] = accounts;
-
+const { loadFixture } = require('@nomicfoundation/hardhat-network-helpers');
+const { deploy, getFactory } = require('../helpers/deploy');
+
+async function fixture() {
+  const accounts = await ethers.getSigners(); // this is slow :/
```

Comment:
This is easily cacheable in [`env.contract.js`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/f92dce51ed24a5b4e0009c1522ee12b30f3409ac/hardhat/env-contract.js)

Reply:
Do you know how this will behave with ethers?
Would it work if in the tests we do `this.accounts.shift()` ?

Reply:
I was thinking in something like:

```js
extendEnvironment(env => {
  const { contract } = env;

  env.contract = function (name, body) {
    const { takeSnapshot } = require('@nomicfoundation/hardhat-network-helpers');

    contract(name, accounts => {
      // reset the state of the chain in between contract test suites
      let snapshot;
      let signers;

      before(async function () {
        signers = signers ?? await ethers.getSigners();
        snapshot = await takeSnapshot();
      });

      after(async function () {
        await snapshot.restore();
      });

      // remove the default account from the accounts list used in tests, in order
      // to protect tests against accidentally passing due to the contract
      // deployer being used subsequently as function caller
      // The signers array is also cloned using the spread operator to avoid changes
      // to the original array.
      body([...signers].slice(1));
    });
  };
});
```

Reply:
And we would get the signers as an argument of `contract("...", accounts => { ... });`
?

That makes sens, but it will also break the existing truffle tests, which we don't want.
Maybe we should do 
```diff
-signers = signers ?? await ethers.getSigners();
+signers = accounts ?? await ethers.getSigners();
```

Reply:
Also, I'm wondering it the `takeSnapshot` / `.restore()` is duplicating effort that `loadFixture` is already doing.
In that case we would keep it until the end of the migration, and remove when there is no longer any truffle tests.

Reply:
Actually, docs confirms that loadFixture takes a snapshot:
https://hardhat.org/hardhat-network-helpers/docs/reference#loadfixture()

Reply:
> And we would get the signers as an argument of contract("...", accounts => { ... });?

Yeah, what do you think of keeping a registry of already migrated contract suites?

```
// env.contract.js
// Remove after completing ethers migration
export const ethersMigratedSuites = {
  Ownable: 'Ownable'
}

...

accounts = ethersMigratedSuites[contract] ? [...signers] : accounts
body(accounts.slice(1));

// Ownable.test.js
describe(ethersMigratedSuites.Ownable, function () { ... }
```

> Also, I'm wondering it the takeSnapshot / .restore() is duplicating effort that loadFixture is already doing.

Yes, it [creates a snapshot](https://github.com/NomicFoundation/hardhat/blob/4bdb96f7dca8e693b710d9f26a6f73d51a9f6ff9/packages/hardhat-network-helpers/src/loadFixture.ts#L62), but I don't think it conflicts with these high-level snapshot.

The reason why we added a snapshot in `env.contract.js` is because we had suspicious that the AccessManager tests were consuming `block.timestamp` by a lot, eventually breaking the deadline in the `ERC2771Forwarder` tests. We couldn't confirm such theory, but if that's the case, we should keep `loadFixture`s per contract suite and also this snapshot.

Reply:
> what do you think of keeping a registry of already migrated contract suites?

Too maintenance heavy (and error prone) IMO.

Reply:
> We couldn't confirm such theory, but if that's the case, we should keep loadFixtures per contract suite and also this snapshot.

If loadFixture does a "per contract suite" snapshot, I don't see why we would have to also do a "cross contract suite" snapshot. That is making everything slower, and I really don't see what it brings in exchange. The only usecase I see for that, is if we wanted our test to all start with `blockNumber = 0` ... and I don't think that is an assumption we should make. IMO our tests (each test.js file) should work regardless of the initial state of the chain.

Reply:
we could do 
```
body(accounts.slice(1), signers.slice(1));
```

And then do 
```
// truffle
contract("...", accounts => { ... });
```
and
```
// ethers
contract("...", (, signers) => { ... });
```

Reply:
Keeping the high-level snapshot makes sense if not all the contract suites implement a fixture. Let's try removing the high-level snapshot once we finish with the migration as well.

> we could do
>
>```
>body(accounts.slice(1), signers.slice(1));
>```

I like this, let's implement it that way 😄